core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java

```diff
@@ -502,6 +507,32 @@ public boolean ownsData() {
       });
    }
 
+   private IntSet computeTransactionOnlySegments(CacheTopology cacheTopology, ConsistentHash newReadCh, Address address) {
+      IntSet transactionOnlySegments = IntSets.mutableEmptySet(numSegments);
+      if (configuration.transaction().lockingMode() == LockingMode.PESSIMISTIC &&
+          cacheTopology.getPendingCH() != null && cacheTopology.getCurrentCH().getMembers().contains(address)) {
+         // In pessimistic caches, the originator does not send the lock command to backups
+         // if it is the primary owner for all the keys.
+         // The idea is that the locks can only be lost completely if the originator crashes (rolling back the tx)
+         // But this means when the owners of a segment change from AB -> BA, B needs to request the transactions
+         // affecting that segment from A, even though B is already an owner of that segment
+         Set<Integer> pendingPrimarySegments = cacheTopology.getPendingCH().getPrimarySegmentsForOwner(address);
+         for (Integer segment : pendingPrimarySegments) {
+            List<Address> currentOwners = newReadCh.locateOwnersForSegment(segment);
```

Comment:
shouldn't check the old read consistent hash?

Reply:
No, I actually meant to write `cacheTopology.getCurrentCH()`.

The only scenario we care about is if a rebalance starts at topology T (NO_REBALANCE) with current CH owners = ABC and installs topology T+1 (READ_OLD_WRITE_ALL) with current CH owners = read owners = ABC and pending CH owners = CBX, write owners = ABCX (with X being either another node D or A).

If A doesn't leave during the rebalance, the primary owner is A then C, and the old read CH is the same as the new read CH for the READ_OLD phase.
If A leaves during the READ_OLD phase, we also get topology T+2 (READ_OLD_WRITE_ALL) with current CH owners = read owners = BC and pending CH owners = CBX, write owners = BCX (X being either D or none).
B can start new transactions in T+2, but because state transfer is ongoing, it will send all lock commands to CX.
If B also leaves, we get topology T+3 (READ_OLD_WRITE_ALL) with current CH owners = read owners = C and pending CH owners = write owners = CX.

So I think it's clearer with `cacheTopology.getCurrentCH()`, but I'll also add a rebalance phase check to make sure we only request transactions during the READ_ONLY_WRITE_ALL phase (actually this should be true for requesting entries as well).